src/functions/ensureEnv.ts

```diff
@@ -0,0 +1,161 @@
+  const checked = await _check(projectId);
+  if (checked) {
+    configstore.set(CONFIGSTORE_KEY, { lastActiveAt: Date.now() });
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Enable the EnvStore API.
+ *
+ * This is an bespoke method of "enabling" EnvStore API. We "enable" the
+ * EnvStore API setting up a non-empty EnvStore ENV_ID=${ENVSTORE_INTERNAL_ID}.
+ */
+export async function enable(projectId: string): Promise<void> {
+  await envstore.patchStore(projectId, ENVSTORE_INTERNAL_ID, { ENABLED: "1" });
+}
+
+interface UserEnv {
+  fnLabel: string;
+  envs: Record<string, string>;
+}
+
+/**
+ * Lookup existing cloud functions and collect user-defined env variables.
+ *
+ * If user's project hasn't activated the EnvStore API, we consider
+ * all non-default environment variable as "user-defined". The user-defined
+ * env is significant because they will be removed on the next deploy following
+ * EnvStore API activation.
+ */
+async function getUserEnvs(projectId: string): Promise<UserEnv[]> {
+  const have = await backend.existingBackend({ projectId, filters: [] }, /* forceRefresh= */ true);
+
+  const fnEnvs: UserEnv[] = [];
+  for (const fn of have.cloudFunctions) {
+    // Filter out non CF3 function instances.
+    if (!deploymentTool.isFirebaseManaged(fn.labels || {})) {
+      continue;
+    }
+    const uenvs: Record<string, string> = {};
+    const envs = fn.environmentVariables;
+    if (envs && Object.keys(envs).length > 1) {
+      // Collect non-default env variables to print.
+      for (const [k, v] of Object.entries(envs)) {
+        if (!DEFAULT_ENV_KEYS.includes(k)) {
+          uenvs[k] = v;
+        }
+      }
+      fnEnvs.push({ fnLabel: helper.getFunctionLabel(fn), envs: uenvs });
+    }
+  }
+  return fnEnvs;
+}
+
+/**
+ * Check if EnvStore API is enabled on the project.
+ * If not enabled, prompt user to enable it.
+ */
+export async function ensure(options: any): Promise<void> {
+  const projectId = getProjectId(options);
+  const isEnabled = await check(projectId);
+  if (isEnabled) {
+    return;
+  }
+
+  logBullet(
+    "functions:env family of commands manages environment variables for your Cloud Functions for Firebase. " +
+      "Learn more about this feature at https://firebase.google.com/docs/functions/env.\n"
+  );
+
+  logWarning(
+    "By opting in to have Firebase manage your function's environment variables, " +
+      "you may lose existing environment variables you've manually set up on your functions. " +
+      "Also, any environment variables you subsequently set with the gcloud command line tool " +
+      "or in the Google Cloud console will be overwritten or removed.\n"
+  );
+
+  const userEnvs = await getUserEnvs(projectId);
+  if (userEnvs.length > 0) {
+    let msg =
+      "If you opt in, the following environment variables will be deleted on next deploy:\n";
+
+    const allKvs: Record<string, string> = {};
+    for (const { fnLabel, envs } of userEnvs) {
```

Comment:
Not asking you to change anything, but I thought this was a fun time to try list comprehensions. If userEnvs was a Record<string, backend.EnvironmentVariables>. It's a fun programming challenge!

```
const functionEnvBreakdown = Object.entries(userEnvs).map(([label, envs]) => {
  const envList = Object.entries(envs).map(([k, v]) => clc.bold(`${k}=${v}`).join(", ");
  return `\t${label}: ${envList}`;
}).join("\n");

const allEnvs = Object.values(userEnvs).reduce((accum, next) => { return {...accum, ...next}}, {});
const allEnvsLabels = Object.entries(allEnvs).map(([k, v]) => `${k}=${v}`).join(" ");
```

This won't work though because of the much bigger issue: We don't handle the edge case where different functions have different values for the same environment variable.

Reply:
Woo these are fun code - going to shamelessly steal them!

As for the last bit - yes it won't detect conflicts, and we don't resolve it, but I'm hoping that it's obvious to the customer that they will have to manually resolve it themselves.